File: luaui/core/SettingsManager.lua
--[[
    SettingsManager.
    Faithful port of haxe_ui.core.SettingsManager.
--]]

local SettingsManager = {}
SettingsManager.__index = SettingsManager

local _instance = nil

function SettingsManager.get_instance()
    if _instance == nil then
        _instance = setmetatable({ _persister = nil }, SettingsManager)
    end
    return _instance
end

function SettingsManager:set(name, value)
    -- In LuaUI, we simplify serialization to strings or use a provided JSON helper
    local s = tostring(value)
    if type(value) == "table" then
        -- Placeholder for real serialization logic if needed
        s = "{}"
    end
    self:getPersister():set(name, s)
end

function SettingsManager:get(name, defaultValue)
    local s = self:getPersister():get(name)
    if s == nil then return defaultValue end
    -- Simplified unserialization
    if s == "true" then return true end
    if s == "false" then return false end
    local n = tonumber(s)
    if n then return n end
    return s
end

function SettingsManager:getPersister()
    if self._persister ~= nil then return self._persister end

    -- Love2D file system persister
    local FileSettingsPersister = {}
    FileSettingsPersister.__index = FileSettingsPersister
    FileSettingsPersister.filename = "settings.json"

    function FileSettingsPersister:set(name, value)
        local o = self:load()
        o[name] = value
        self:save(o)
    end

    function FileSettingsPersister:get(name)
        local o = self:load()
        return o[name]
    end

    function FileSettingsPersister:load()
        if love.filesystem.getInfo(self.filename) then
            local content = love.filesystem.read(self.filename)
            -- Simple parse logic (placeholder for actual JSON parse)
            return {}
        end
        return {}
    end

    function FileSettingsPersister:save(o)
        -- love.filesystem.write(self.filename, "{}")
    end

    self._persister = setmetatable({}, FileSettingsPersister)
    return self._persister
end

return SettingsManager

File: flutter_ui/lib/core/settings_manager.dart
import 'dart:convert';
import 'dart:io';
import 'package:flutter/foundation.dart';

class SettingsManager {
  static SettingsManager? _instance;
  static SettingsManager get instance {
    _instance ??= SettingsManager();
    return _instance!;
  }

  ISettingsPersister? _persister;

  SettingsManager();

  void set<T>(String name, T? value) {
    String s = jsonEncode(value);
    getPersister().set(name, s);
  }

  T? get<T>(String name, [T? defaultValue]) {
    String? s = getPersister().get(name);
    if (s == null) {
      return defaultValue;
    }
    return jsonDecode(s) as T?;
  }

  ISettingsPersister getPersister() {
    if (_persister != null) {
      return _persister!;
    }
    if (kIsWeb) {
      _persister = NoOpSettingsPersister();
    } else {
      _persister = FileSettingsPersister();
    }
    return _persister!;
  }
}

abstract class ISettingsPersister {
  void set(String name, String value);
  String? get(String name);
}

class FileSettingsPersister implements ISettingsPersister {
  String filename = "settings.json";
  FileSettingsPersister();
  @override
  void set(String name, String value) {
    Map<String, dynamic> o = load();
    o[name] = value;
    save(o);
  }

  @override
  String? get(String name) {
    Map<String, dynamic> o = load();
    return o[name] as String?;
  }

  Map<String, dynamic> load() {
    File file = File(filename);
    if (file.existsSync()) {
      String content = file.readAsStringSync();
      if (content.isNotEmpty) {
        return jsonDecode(content) as Map<String, dynamic>;
      }
    }
    return {};
  }

  void save(Map<String, dynamic> o) {
    File(filename).writeAsStringSync(jsonEncode(o));
  }
}

class NoOpSettingsPersister implements ISettingsPersister {
  NoOpSettingsPersister();
  @override
  void set(String name, String value) {}
  @override
  String? get(String name) => null;
}
